### ISSUE TITLE
storage: investigate bug exposing multiple intents in interleavingIntentsIter

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#86546]
See: #86441

Jira issue: CRDB-18808

**Comment by tbg** [comment]
I looked into this and I think the bug is with #86441, not intentInterleavingIter. The reduced test case I have been looking at (on top of the code in #86441) is this:

<details>
<summary>...</summary>

```

# Perform some writes at various sequence numbers.

run ok
with t=A
  txn_begin ts=11
  txn_step  seq=10
  put       k=k  v=a
  txn_step  seq=20
  put       k=k  v=b
  txn_step  seq=30
  put       k=k  v=c
  txn_step  seq=40
----
>> at end:
txn: "A" meta={id=00000000 key=/Min pri=0.00000000 epo=0 ts=11.000000000,0 min=0,0 seq=40} lock=true stat=PENDING rts=11.000000000,0 wto=false gul=0,0
meta: "k"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=11.000000000,0 min=0,0 seq=30} ts=11.000000000,0 del=false klen=12 vlen=6 ih={{10 /BYTES/a}{20 /BYTES/b}} mergeTs=<nil> txnDidNotUpdateMeta=false
data: "k"/11.000000000,0 -> /BYTES/c

# Mask a single write.
# The third write should be hidden now.

run ok
with t=A
  txn_ignore_seqs seqs=(25-35)
  scan k=k end=-k
----
scan: "k" -> /BYTES/b @11.000000000,0
scan: "k" -> /BYTES/c @11.000000000,0
>> at end:
txn: "A" meta={id=00000000 key=/Min pri=0.00000000 epo=0 ts=11.000000000,0 min=0,0 seq=40} lock=true stat=PENDING rts=11.000000000,0 wto=false gul=0,0 isn=1

```

</details>

so basically you have an intent that stores an intent history that has `b` at seq 20 (and `a` at 10 though that doesn't matter) and then the next key is the version key for the most recent write `c`. But we are going to scan in a way that ignores `c`, seeing `b` instead.

When we scan, we first find the intent. `pebbleMVCCScanner` realizes that we're reading from the intent history, so it produces `b` as the output.

This has the side effect of updating `p.curUnsafeKey.Timestamp` to the timestamp of that intent (`11`):

https://github.com/cockroachdb/cockroach/blob/9c7db33c1ee21145fe372d61ad26b45a9f151ee3/pkg/storage/pebble_mvcc_scanner.go#L822-L831

Then we advance the parent iterator, which now produces `c`, as it ought to (as that is the key next to the intent). But that's also a version at timestamp `11` so now we're throwing off the code in #86441 which now gets tricked into thinking that it's already on the next MVCC key. That then leads to `c` being emitted as well, which was not the plan.

The way to avoid this problem is to avoid corrupting `unsafeKey`, i.e. passing the timestamp at which the version is to be emitted into `addAndAdvance` explicitly. Then, `nextKey` will start out with `prevTS=0` (from the intent metadata) and the short-circuit will be avoided as we are not visiting the physical MVCC history without mixing it up with the IntentHistory.

I'll give that a shot.

**Comment by erikgrinaker** [comment]
Thanks for looking into this! I figured it was something like this. But I don't think we're going to do anything about the optimization over in #85610, given stability. @jbowens is also doing some work over in https://github.com/cockroachdb/pebble/pull/1860 to push this down into Pebble.

**Comment by tbg** [comment]
Partial WIP to fix up #86441 in https://github.com/cockroachdb/cockroach/pull/86568, but abandoning it.
